azurelinuxagent/common/protocol/goal_state.py

```diff
@@ -292,11 +292,31 @@ def _update(self, force_update):
         #
         if self._extensions_goal_state.source == GoalStateSource.FastTrack and self._goal_state_properties & GoalStateProperties.Certificates:
             self._check_certificates()
+            # Re-download certificates if any have been removed from disk since last download
+            if self._certs_uri is not None:
+                certificates = self.certs.summary
+                certs_missing_from_disk = False
+
+                for c in certificates:
+                    cert_path = os.path.join(conf.get_lib_dir(), c['thumbprint'] + '.crt')
+                    if not os.path.isfile(cert_path):
+                        certs_missing_from_disk = True
+                        message = "Certificate required by goal state is not on disk: (0)".format(cert_path)
+                        self.logger.info(message)
+                        add_event(op=WALAEventOperation.GoalState, message=message)
+                if certs_missing_from_disk:
+                    # Try to re-download certs. Sometimes download may fail if certs_uri is outdated/contains wrong
+                    # container id (for example, when the VM is moved to a new container after resuming from
+                    # hibernation). If download fails we should report and continue with goal state processing, as some
+                    # extensions in the goal state may succeed.
+                    try:
+                        self._download_certificates(self._certs_uri)
```

Comment:
this PR is trying to handle that case. When we resume from hibernation we don't force an update of the goal state/refresh the cert_uri, so after we added logic to download certs on every fasttrack GS, the download failed because cert_uri was outdated. When download failed we didn't handle the exception, this PR handles the exception in this case by reporting the failure and continuing. We also are trying to limit the # of times we hit this issue by only downloading certs when they're missing from the disk.

Cert_uri is only refreshed if we force an update or fetch new GS

Reply:
The goal state is refreshed by the code handling the InconsistentGoalState exception. The change to force download on every FT goal state bypassed that logic. This changes restores that path by downloading the certs only after the check for consistency has succeeded (plus includes some other improvements, like better error handling, logging and telemetry)

Reply:
Thanks for explaining.